rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileDownloader.java

```diff
@@ -0,0 +1,120 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * Created by Abhishek on 3/9/18.
+ */
+class FileDownloader
+{
+  Configuration conf = null;
+  private ExecutorService processService;
+  int diskReadBufferSize;
+
+  private static final Log log = LogFactory.getLog(FileDownloader.class);
+  private static DirectBufferPool bufferPool = new DirectBufferPool();
+
+  public FileDownloader(Configuration conf)
+  {
+    this.conf = conf;
+    int numThreads = CacheConfig.getRemoteFetchNumThreads(conf);
+    this.diskReadBufferSize = CacheConfig.getDiskReadBufferSizeDefault(conf);
+
+    ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(numThreads);
+    processService = MoreExecutors.getExitingExecutorService(executor);
+  }
+
+  protected List<FileDownloadRequestChain> getFileDownloadRequestChains(ConcurrentMap<String, DownloadRequestContext> contextMap)
+      throws IOException
+  {
+    List<FileDownloadRequestChain> readRequestChainList = new ArrayList<FileDownloadRequestChain>();
+    for (Map.Entry<String, DownloadRequestContext> entry : contextMap.entrySet()) {
+      Path path = new Path(entry.getKey());
+      DownloadRequestContext context = entry.getValue();
+
+      FileSystem fs = path.getFileSystem(conf);
+      fs.initialize(path.toUri(), conf);
+      String localPath = CacheConfig.getLocalPath(entry.getKey(), conf);
+      log.info("Processing Request for File : " + path.toString() + " LocalFile : " + localPath);
+      ByteBuffer directWriteBuffer = bufferPool.getBuffer(diskReadBufferSize);
+
+      FileDownloadRequestChain requestChain = new FileDownloadRequestChain(fs, localPath,
+          directWriteBuffer, conf, context.getRemoteFilePath(), context.getFileSize(),
+          context.getLastModifiedTime());
+
+      for (Range<Long> range : context.getRanges().asRanges()) {
+        ReadRequest request = new ReadRequest(range.lowerEndpoint(), range.upperEndpoint(),
+            range.lowerEndpoint(), range.upperEndpoint(), null, 0, context.getFileSize());
+        requestChain.addReadRequest(request);
+      }
+
+      log.debug("Request added for file: " + requestChain.getRemotePath() + " Number of Requests : " +
+          requestChain.getReadRequests().size());
+
+      readRequestChainList.add(requestChain);
+    }
+
+    return readRequestChainList;
+  }
+
+  protected void processDownloadRequests(List<FileDownloadRequestChain> readRequestChainList)
+  {
+    int sizeRead = 0;
+    List<Future<Integer>> futures = new ArrayList<Future<Integer>>();
+
+    for (FileDownloadRequestChain requestChain : readRequestChainList) {
+      requestChain.lock();
+      Future<Integer> result = processService.submit(requestChain);
```

Comment:
We need to think about de-duping requests here as well. There is a high chance that this executor is overwhelmed and the queue contains lot of repeated DownloadRequest-ranges. If not doing in this PR, please open a issue to track it

Reply:
Yes we will raise another PR as an enhancement for this